### PR TITLE
PCHR-782: Dashboard (Fixes)

### PR DESCRIPTION
**Pull request by AkA84**
This PR introduces various fixes and improvements to the UI, frontend code and backend code

##### Before
first row - before
##### After
first row - after

##### Before
steps - before
##### After
steps - after

#### 3. Form errors
form errors - after

#### 4. Loading spinner
spinner - after

#### 5. Due dates warning on edit
due dates warning

##### Before
order - before
##### After
order - after

#### 7. Active cycle checkbox
active checkbox - after

#### 8. "No cycles found" message
no cycles found msg - after

---

#### 1. Form validation
form validation - after
#### 2. Active cycle count
active count - after
#### 3. Total appraisals number in all cycles
total appraisals count - after
#### 4. Blue heading count
blue heading - after
#### 5. Dynamic data on appraisal elements
appraisals elements - after

### CODE REVIEW

File: uk.co.compucorp.civicrm.appraisals/CRM/Appraisals/BAO/AppraisalCycle.php [hunk 1]
```diff
@@ -240,33 +241,106 @@ public static function getStatusOverview($managerId) {
...
+     *       contain a placeholder %paramIndex_<index> (i.e.: $paramIndex_1)
+     *       that will be used by CRM_Core_DAO::executeQuery to fill in the real value
+     *     values - An array of value that CRM_Core_DAO::executeQuery will
+     *       use in the final query
+     *     type - The type of the values used by CRM_Core_DAO::executeQuery
+     */
+    private function fillQueryPlaceholder(&$query, &$params, $options) {
+        $replacement = '';
+
+        if (!empty(array_filter($options['values']))) {
```
Comment (Agares4) [inline uk.co.compucorp.civicrm.appraisals/CRM/Appraisals/BAO/AppraisalCycle.php]: why array_filter here, if later you don't care about empty values in the array?

File: uk.co.compucorp.civicrm.appraisals/CRM/Appraisals/BAO/AppraisalCycle.php [hunk 2]
```diff
@@ -240,33 +241,106 @@ public static function getStatusOverview($managerId) {
...
+            $replacement = $options['string'];
+
+            foreach ($options['values'] as $key => $value) {
+                $paramIndex = count($params) + 1;
+                $params[$paramIndex] = array($value, $options['type']);
```
Comment (Agares4) [inline uk.co.compucorp.civicrm.appraisals/CRM/Appraisals/BAO/AppraisalCycle.php]: If you just want to add to the end of array, you can use: `$params[] = array(...)`

File: uk.co.compucorp.civicrm.appraisals/CRM/Appraisals/BAO/AppraisalCycle.php (hunk 1 above)
Comment (AkA84) [inline uk.co.compucorp.civicrm.appraisals/CRM/Appraisals/BAO/AppraisalCycle.php]: Sorry i'm not sure I understand the question

**Comment by kkalamarski** [comment]
What does the asterix symbol (*) mean in point 7. ? 
Does it mean that it must be checked - cycle have to be active in order to add it, 
or it's a [null, true, false] checkbox and need either to be checked or unchecked?

File: uk.co.compucorp.civicrm.appraisals/CRM/Appraisals/BAO/AppraisalCycle.php (hunk 1 above)
Comment (Agares4) [inline uk.co.compucorp.civicrm.appraisals/CRM/Appraisals/BAO/AppraisalCycle.php]: You use array_filter here to remove 0s, empty strings, nulls, etc. from $option['values'] and check if that's not empty. But later on you just iterate over the whole array.

File: uk.co.compucorp.civicrm.appraisals/js/src/appraisals/controllers/appraisals-dashboard-ctrl.js [hunk 3]
```diff
@@ -74,49 +75,56 @@ define([
...
-                            vm.cycles.push(cycle);
-                        });
-                    } else {
-                        vm.cycles = cycles.list;
-                    }
-
-                    vm.loadingDone = vm.cycles.length === cycles.total;
-                });
+                vm.loading.inProgress = true;
+                vm.pagination.page = !!addPage ? vm.pagination.page + 1 : 1;
```
Comment (kkalamarski) [inline uk.co.compucorp.civicrm.appraisals/js/src/appraisals/controllers/appraisals-dashboard-ctrl.js]: If it's a boolean anyway is the double negation '!!' needed?

Comment (AkA84) [inline uk.co.compucorp.civicrm.appraisals/js/src/appraisals/controllers/appraisals-dashboard-ctrl.js]: Well not strictly, but given that it's a parameter passed to the function i think it's better to always force it to be a boolean

**Comment by AkA84** [comment]
> What does the asterix symbol (*) mean in point 7. ? 

Well it means "mandatory", but given you can submit the form with the checkbox either checked or unchecked, I realize now it doesn't make much sense. I'll remove it

File: uk.co.compucorp.civicrm.appraisals/js/src/appraisals/controllers/appraisals-dashboard-ctrl.js (hunk 3 above)
Comment (kkalamarski) [inline uk.co.compucorp.civicrm.appraisals/js/src/appraisals/controllers/appraisals-dashboard-ctrl.js]: Okay, makes sense.

**Comment by kkalamarski** [comment]
Approved

**Pull request merged**